src/Features/CustomerEffortScoreTracks.php

```diff
@@ -0,0 +1,146 @@
+<?php
+/**
+ * WooCommerce Customer effort score tracks
+ *
+ * @package WooCommerce\Admin\Features
+ */
+
+namespace Automattic\WooCommerce\Admin\Features;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Triggers customer effort score on several different actions.
+ */
+class CustomerEffortScoreTracks {
+	const CES_TRACKS_QUEUE_OPTION_NAME       = 'woocommerce_ces_tracks_queue';
+	const CLEAR_CES_TRACKS_QUEUE_OPTION_NAME = 'woocommerce_clear_ces_tracks_queue';
+
+	/**
+	 * Constructor. Sets up filters to hook into WooCommerce.
+	 */
+	public function __construct() {
+		add_action(
+			'transition_post_status',
+			array(
+				$this,
+				'run_on_transition_post_status',
+			),
+			10,
+			3
+		);
+		add_action(
+			'admin_init',
+			array(
+				$this,
+				'maybe_clear_ces_tracks_queue',
+			)
+		);
+	}
+
+	/**
+	 * Hook into the post status lifecycle, only interested in products that
+	 * are either being added or edited.
+	 *
+	 * @param string $new_status The new status.
+	 * @param string $old_status The old status.
+	 * @param Post   $post       The post.
+	 */
+	public function run_on_transition_post_status(
+		$new_status,
+		$old_status,
+		$post
+	) {
+		if ( 'product' !== $post->post_type ) {
+			return;
+		}
+
+		if ( 'publish' !== $new_status ) {
+			return;
+		}
+
+		if ( 'publish' !== $old_status ) {
+			$this->enqueue_ces_survey_for_new_product();
+		} else {
+			$this->enqueue_ces_survey_for_edited_product();
+		}
+	}
+
+	/**
+	 * Get the current published product count.
+	 *
+	 * @return integer The current published product count.
+	 */
+	private function get_product_count() {
+		$query         = new \WC_Product_Query(
+			array(
+				'limit'    => 1,
+				'paginate' => true,
+				'return'   => 'ids',
+				'status'   => array( 'publish' ),
+			)
+		);
+		$products      = $query->get_products();
+		$product_count = intval( $products->total );
+
+		return $product_count;
+	}
+
+	/**
+	 * Enqueue the CES survey trigger for a new product.
+	 */
+	private function enqueue_ces_survey_for_new_product() {
+		$queue = get_option( self::CES_TRACKS_QUEUE_OPTION_NAME, array() );
```

Comment:
Once we're past the "spike" version of this, some helper methods to manage the queue would probably be helpful.